README_tcpdevmem.md

```diff
@@ -0,0 +1,277 @@
+    - [Override CUDA library directory (DLVM)](#override-cuda-library-directory-dlvm)
+  - [TCPDirect UDMA: Compiling tcp\_stream](#tcpdirect-udma-compiling-tcp_stream)
+  - [Running tcp\_stream](#running-tcp_stream)
+    - [Added flags](#added-flags)
+    - [Running tcp\_stream via `multi_neper.py`](#running-tcp_stream-via-multi_neperpy)
+      - [Example of successful output](#example-of-successful-output)
+    - [Running tcp\_stream directly](#running-tcp_stream-directly)
+
+
+## TCPDirect CUDA: tcp_stream within Docker container
+
+A TCPDirect-Cuda enabled Docker image exists at `gcr.io/a3-tcpd-staging-hostpool/neper`. The `run_neper_container.sh` script makes it easy to run Neper:
+
+
+```
+./run_neper_container.sh
+
+# within the container
+FLOWS=2
+BUF_SIZE=409600
+DEVS=eth1,eth2,eth3,eth4
+DSTS=192.168.1.26,192.168.2.26,192.168.3.26,192.168.4.26
+SRCS=192.168.1.23,192.168.2.23,192.168.3.23,192.168.4.23
+./multi_neper.py --hosts $DSTS \
+  --devices $DEVS --buffer-size $BUF_SIZE \
+  --flows $FLOWS --threads $FLOWS \
+  --src-ips $SRCS --log DEBUG \
+  --q-num $FLOWS --phys-len 2147483648 \
+  --client \
+  --mode cuda
+```
+
+
+
+### Building your own image for testing
+
+
+```
+# within the repo directory
+docker build -t neperc .
+
+function run_neper_container() {
+  docker run \
+    --name neper_c \
+    -u 0 --network=host \
+    --cap-add=IPC_LOCK \
+    --userns=host \
+    --volume /run/tcpx:/tmp \
+    --volume /var/lib/nvidia/lib64:/usr/local/nvidia/lib64 \
+    --volume /var/lib/tcpx:/usr/local/tcpx \
+    --shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864 \
+    --device /dev/nvidia0:/dev/nvidia0 \
+    --device /dev/nvidia1:/dev/nvidia1 \
+    --device /dev/nvidia2:/dev/nvidia2 \
+    --device /dev/nvidia3:/dev/nvidia3 \
+    --device /dev/nvidia4:/dev/nvidia4 \
+    --device /dev/nvidia5:/dev/nvidia5 \
+    --device /dev/nvidia6:/dev/nvidia6 \
+    --device /dev/nvidia7:/dev/nvidia7 \
+    --device /dev/nvidia-uvm:/dev/nvidia-uvm \
+    --device /dev/nvidiactl:/dev/nvidiactl \
+    --cap-add=NET_ADMIN \
+    --env LD_LIBRARY_PATH=/usr/local/nvidia/lib64:/usr/local/tcpx/lib64 \
+    "$@"
+}
+
+run_neper_container -it neperc bash
+```
+
+You can replace the `neperc` image name and run `docker push ${image_name}` if you’re building the image on one host, and running a container on another.
+
+
+### Override CUDA library directory (DLVM)
+
+The script assumes that `libcuda.so*` files are found in `/var/lib/nvidia/lib64`. In case this isn’t true (like when on DLVM), you can override the default env var: `CUDA_LIB_DIR`:
+
+```
+CUDA_LIB_DIR=/usr/lib/x86_64-linux-gnu ./run_neper_container.sh
+```
+
+
+## TCPDirect UDMA: Compiling tcp_stream
+
+Neper can be built statically on a host with UDMA header files.
+
+The Makefile assumes that the header files are found at `~/cos-kernel/usr/include`. You can manually change the directory in the Makefile:
```

Comment:
No longer true, right? Buliding should not require manually changing the directory in the makefile now, I think.

Reply:
removed